sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala

```diff
@@ -0,0 +1,123 @@
+    traverseAndSubstituteCTE(plan, false)
+  }
+
+  private def legacyTraverseAndSubstituteCTE(plan: LogicalPlan): LogicalPlan = {
+    plan.resolveOperatorsUp {
+      case With(child, relations) =>
+        // substitute CTE expressions right-to-left to resolve references to previous CTEs:
+        // with a as (select * from t), b as (select * from a) select * from b
+        relations.foldRight(child) {
+          case ((cteName, ctePlan), currentPlan) => substituteCTE(currentPlan, cteName, ctePlan)
+        }
+    }
+  }
+
+  /**
+   * Traverse the plan and expression nodes as a tree and replace matching references to CTE
+   * definitions.
+   * - If the rule encounters a WITH node then it substitutes the child of the node with CTE
+   *   definitions of the node right-to-left order as a definition can reference to a previous
+   *   one.
+   *   For example the following query is valid:
+   *   WITH
+   *     t AS (SELECT 1),
+   *     t2 AS (SELECT * FROM t)
+   *   SELECT * FROM t2
+   * - If a CTE definition contains an inner WITH node then substitution of inner should take
+   *   precedence because it can shadow an outer CTE definition.
+   *   For example the following query should return 2:
+   *   WITH
+   *     t AS (SELECT 1),
+   *     t2 AS (
+   *       WITH t AS (SELECT 2)
+   *       SELECT * FROM t
+   *     )
+   *   SELECT * FROM t2
+   * - If a CTE definition contains a subquery that contains an inner WITH node then substitution
+   *   of inner should take precedence because it can shadow an outer CTE definition.
+   *   For example the following query should return 2:
+   *   WITH t AS (SELECT 1 AS c)
+   *   SELECT max(c) FROM (
+   *     WITH t AS (SELECT 2 AS c)
+   *     SELECT * FROM t
+   *   )
+   * - If a CTE definition contains a subquery expression that contains an inner WITH node then
+   *   substitution of inner should take precedence because it can shadow an outer CTE
+   *   definition.
+   *   For example the following query should return 2:
+   *   WITH t AS (SELECT 1)
+   *   SELECT (
+   *     WITH t AS (SELECT 2)
+   *     SELECT * FROM t
+   *   )
+   * @param plan  the plan to be traversed
+   * @param inTraverse whether the current traverse is called from another traverse, only in this
+   *                   case name collision can occur
+   * @return then plan where CTE substitution is applied
+   */
+  private def traverseAndSubstituteCTE(plan: LogicalPlan, inTraverse: Boolean): LogicalPlan = {
+    plan.resolveOperatorsUp {
+      case With(child: LogicalPlan, relations) =>
+        val traversedChild = child transformExpressions {
+          case e: SubqueryExpression => e.withNewPlan(traverseAndSubstituteCTE(e.plan, true))
+        }
+
+        relations.foldRight(traversedChild) {
+          case ((cteName, ctePlan), currentPlan) =>
+            lazy val substitutedCTEPlan = traverseAndSubstituteCTE(ctePlan, true)
+            substituteCTE(currentPlan, cteName, substitutedCTEPlan)
```

Comment:
It seems this line has an issue about type inferences... could you check again?

Reply:
This is a bit strange, because it compiles, but my InteliJ also complains about it. If I force `val traversedChild: LogicalPlan` the issue disappears.